bson/bsonrw/extjson_reader.go

```diff
@@ -159,26 +159,18 @@ func (ejvr *extJSONValueReader) pop() {
 	}
 }
 
-func (ejvr *extJSONValueReader) skipDocument() error {
-	// read entire document until ErrEOD (using readKey and readValue)
-	_, typ, err := ejvr.p.readKey()
-	for err == nil {
-		_, err = ejvr.p.readValue(typ)
-		if err != nil {
-			break
-		}
-
-		_, typ, err = ejvr.p.readKey()
-	}
-
-	return err
-}
-
-func (ejvr *extJSONValueReader) skipArray() error {
-	// read entire array until ErrEOA (using peekType)
+func (ejvr *extJSONValueReader) skipObject() error {
```

Comment:
I think we can tweak this to only use `advanceState`.

The initial state, `ejvr.p.s` should to be `jpsSawBeginArray` or `jpsSawBeginObject` at the beginning of this function. I think we can do something like:

- initialize `count` to 1
- call advanceState
- if the state is `jpsSawBeginArray` or `jpsSawBeginObject`, increment `count`
- if the state is `jpsSawEndArray` or `jpsSawEndObject`, decrement `count`
- repeat until `count` is 0

Reply:
Sounds good! That algorithm seems to work perfectly for all existing examples.